ENH: Tie weights for target_modules in Lora (#2864)

romitjain opened the pull request:
Solves #2864 for `target_modules`

Enables `ensure_weight_tying` flag in `LoraConfig` for `target_modules`.

For LoRA, if any of the tied layers are added to `target_modules` and `ensure_weight_tying == True`, the adapters added to the layer are shared with all the tied layers.

For example, if a model has tied weights and `target_modules=['embed_tokens']` then, LoRA adapters are added to both `embed_tokens` and `lm_head`. The adapters in `lm_head` share the weights with the adapters added to `embed_tokens`

BenjaminBossan commented on src/peft/tuners/lora/config.py:
```diff
@@ -382,6 +382,11 @@ class LoraConfig(PeftConfig):
             `target_parameters`. As an example, for Llama4, you can pass:
             `target_parameters=['feed_forward.experts.gate_up_proj', 'feed_forward.experts.down_proj]`. Passing a
             string for regex matching is not implemented yet.
+        ensure_weight_tying (`bool`, *optional*)
```
Thanks for adding.

BenjaminBossan reviewed (commented):
Thanks for this draft PR to extend the feature to `target_modules`. I haven't done a full review yet, as some implementation details have yet to be figured out, but I gave some early feedback. This feature could be a bit more difficult to implement than for `modules_to_save`, I added some comments on why, please check.

romitjain commented on src/peft/tuners/tuners_utils.py:
```diff
@@ -910,15 +950,6 @@ def inject_adapter(
                     RuntimeWarning,
                 )
 
-        tied_target_modules = self._get_tied_target_modules(model=model)
```
This is removed since this is no longer needed. We have fixed this warning

romitjain commented:
@BenjaminBossan This is now ready for review. I have also updated the logic for tied layers in `modules_to_save` so that `lm_head` and `[embed_tokens, lm_head]` cases are supported. Earlier, they would not have worked. The high level implementation remains the same but according to me it's much better placed then my earlier commits.

I have also added a few tests for the above case, and all of the tests pass.

The only thing remaining is how to check for `target_modules` in case it's a string. I will come back to it, but you can go ahead and review the core logic.

BenjaminBossan commented on src/peft/tuners/lora/layer.py:
```diff
@@ -195,6 +197,16 @@ def update_layer(
+            lora_B_params = tied_adapters["lora_B"]
+
+            self.lora_A[adapter_name].weight = torch.nn.Parameter(lora_A_params)
+            self.lora_B[adapter_name].weight = torch.nn.Parameter(lora_B_params)
```
Let's add a comment here that weight tying is only implemented for `Embedding` => `Linear`, which is why this code is not necessary to be added to the other layer types (say, the conv layers).

BenjaminBossan commented on src/peft/tuners/tuners_utils.py:
```diff
@@ -805,6 +833,9 @@ def inject_adapter(
                 if key not in module_names:
                     unmatched_modules.append(key)
                 else:
+                    if self._check_tied_module_exists(peft_config, key):
```
Let's add a comment here why this is skipped for now.

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -4839,69 +4839,71 @@ def get_input_embeddings(self):
         return CausalLM().eval().to(self.torch_device)
 
-    def test_weight_tying_tied_model_lora(self):
+    @pytest.mark.parametrize("layer", ["lm_head", "embed_tokens", ["lm_head", "embed_tokens"]])
```
```suggestion
    @pytest.mark.parametrize("modules_to_save", [["lm_head"], ["embed_tokens"], ["lm_head", "embed_tokens"]])
```

Let's call this `modules_to_save` to make it immediately clear what is meant and also put everything into a list, so that we don't need to check `if isinstance(layer, list)` below. Same for the other tests.

githubnemo replied:
Part 1 of the comment is not resolved. I think all these tests would benefit from `s/layer/modules_to_save/` here.

BenjaminBossan reviewed (changes requested):
Thanks for the latest updates. This looks much cleaner now, I think we're approaching the finish line. There were still some issues I had though, so please check my comments.

BenjaminBossan commented on src/peft/peft_model.py:
```diff
@@ -315,6 +315,11 @@ def save_mutated_as_lora(peft_config, path_initial_model_for_weight_conversion,
                         peft_config, path_initial_model_for_weight_conversion, output_state_dict, kwargs
                     )
+
+                for k, v in output_state_dict.items():
```
Why was this added?

BenjaminBossan replied:
Could we make the tensor contiguous right there and remove the call here?

BenjaminBossan commented on src/peft/tuners/lora/model.py:
```diff
@@ -249,7 +261,8 @@ def _create_and_replace(
                 new_module.requires_grad_(False)
-            self._replace_module(parent, target_name, new_module, target)
+
+            self._replace_module(parent=parent, child_name=target_name, new_module=new_module, child=target)
```
Although I'm not against using keywords here, this method is called numerous times in PEFT with just positional args, so let's leave it at that for consistency.

BenjaminBossan commented on src/peft/tuners/lora/model.py:
```diff
@@ -806,8 +819,45 @@ def subtract_mutated_init(self, output_state_dict: dict[str, torch.Tensor], adap
+            tied_weight_keys (list[str])
+        """
+        tied_weight_keys = set(tied_weight_keys)
+        setattr(peft_config, "modules_to_tie", tied_weight_keys)
```
```suggestion
        peft_config.modules_to_tie = tied_weight_keys
```

BenjaminBossan commented on src/peft/tuners/lora/model.py:
```diff
@@ -806,8 +819,45 @@ def subtract_mutated_init(self, output_state_dict: dict[str, torch.Tensor], adap
+            if m in modules_to_save:
+                modules_to_save.remove(m)
+
+        setattr(peft_config, "modules_to_save", modules_to_save)
```
```suggestion
        peft_config.modules_to_save = modules_to_save
```

BenjaminBossan commented on src/peft/tuners/lora/model.py:
```diff
@@ -811,3 +826,11 @@ def _add_modules_to_tie(self, peft_config, tied_weight_keys):
         peft_config.modules_to_tie = missing_keys
+
+    def _add_targets_to_tie(self, peft_config, tied_weight_keys):
+        target_modules = set(getattr(peft_config, "target_modules", []) or [])
```
Sounds good. This has yet to be updated, right?

romitjain replied:
Yes, I will do this in the next commit

romitjain commented:
@BenjaminBossan please review the latest changes now. I believe I have addressed all your comments, but let me know if I missed something.

I have added test cases where we are passing `target_modules` as `str` and added a (slightly) hacky solve for that. It's opinionated to keep the flow simple. 

Regarding the transformers v5 update, since we would be having a version locked in peft, I believe if this PR advances faster than that, we can merge this. I can take up changes too whenever they're needed. However, you are much closer to this, so you can decide and let me know.

BenjaminBossan commented on src/peft/tuners/tuners_utils.py:
```diff
@@ -363,6 +363,30 @@ def _prepare_model(self, peft_config: PeftConfig, model: nn.Module):
+        ):
+            return True
+
+        return False
```
```suggestion
        target_modules_to_tie = getattr(peft_config, "target_modules_to_tie", []) or []
        return key in target_modules_to_tie or any(
            key.endswith(f".{target_key}") for target_key in target_modules_to_tie
        )
```

Reason:
1. More concise
2. Doesn't really make sense to use dicts (`{}`) as default. If you meant to use the empty set, I don't think it amortizes here.

romitjain commented on src/peft/tuners/lora/model.py:
```diff
@@ -806,8 +819,56 @@ def subtract_mutated_init(self, output_state_dict: dict[str, torch.Tensor], adap
+        peft_config.modules_to_tie = tied_weight_keys
+
+        modules_to_save = getattr(peft_config, "modules_to_save", []) or []
+        if "embed_tokens" not in modules_to_save:
```
Yes, that is true. I am not able to find a way to get the embedding layer name from the model

BenjaminBossan replied:
We could try to find the layer name whose parameter corresponds to `model.get_input_embedding()` but I'm fine with assuming the name here. Let's just add a comment.

romitjain commented on src/peft/tuners/lora/model.py:
```diff
@@ -806,8 +819,56 @@ def subtract_mutated_init(self, output_state_dict: dict[str, torch.Tensor], adap
+        tied_weight_keys = set(tied_weight_keys)
+        peft_config.target_modules_to_tie = tied_weight_keys
+
+        raw_target_modules = getattr(peft_config, "target_modules", None)
```
I redid this a bit. We just need to make sure that `embed_tokens` is present in the `target_modules`

romitjain commented:
@BenjaminBossan I have made the updates. As of now, we have 3 outstanding updates we need to resolve

1. What to do in case `target_modules` is a str: https://github.com/huggingface/peft/pull/2879#discussion_r2503550670: I have replied to the comment with a small modification in the logic
2. Release of this PR: Can this PR be merged to main but not in release? Or do we have an ETA for v5 arrival? Some of our internal PRs depend on this, hence asking.
3. (Minor) Duplication of a warning message: https://github.com/huggingface/peft/pull/2879#discussion_r2510738148L I have replied to the comment

BenjaminBossan commented on src/peft/tuners/tuners_utils.py:
```diff
@@ -1198,6 +1238,24 @@ def _add_modules_to_tie(self, peft_config, tied_weight_keys):
+        Check `peft.tuners.lora.LoraModel._add_modules_to_tie` for an example.
+        """
+        msg = (
+            "Model has `tie_word_embeddings=True` and a tied layer is part of the adapter, "
```
We can just make it a global, I don't see an issue with that.

romitjain commented:
1. https://github.com/huggingface/peft/pull/2879#discussion_r2518948684: Good suggestion. I have added a new function to find a layer by reference tensors
2. I have resolved the merge conflict and addressed your remaining comments
3. Here's a psudo-code of the complete flow that is currently implemented
```
def inject_adapter():
    ...

    # finds tied modules and adds to a set - target_modules_to_tie
    self._check_tied_modules(model, peft_config)

    tied_targets = []

    # loop 1
    for k in keys:
        result = k in target_modules
        to_tie = k in target_modules_to_tie

        if to_tie:
            tied_targets.append(k)
            continue
        if result:
            add_lora(k)

    # loop 2
    for t in tied_targets:
        add_tied_lora(t)

    ...
```
4. Here is the flow of the alternate that you suggested

```
def inject_adapter():
    ...
    # loop 1
    for k in keys:
        result = k in target_modules
        if result:
            add_lora(k)


    # finds tied modules and adds to a set - target_modules_to_tie
    self._check_tied_modules(model, peft_config)

    # this is needed since all the tied adapters reference embedding
    # layer's lora as source
    add_lora(embed_tokens)

    # loop 2
    for t in target_modules_to_tie:
        remove_lora(t) # will remove lora if it exists, this might be a bit complex
        add_tied_lora(t)

    ...
```

BenjaminBossan reviewed (changes requested):
Thanks for the latest improvements to the PR. As you can see, there is one test failing because of a warning not matching, could you please take a look?

> Here's a psudo-code of the complete flow that is currently implemented

Thanks for this. I think the second approach wouldn't actually differ that much when it comes to complexity, as such I'm fine with keeping the current approach.

romitjain commented:
@BenjaminBossan Resolved your comment

romitjain commented:
Hi @githubnemo, it would be very helpful if you could review the PR. One of our internal features depends on this :)

romitjain commented:
Hi @githubnemo, gentle reminder. Would really appreciate an update. Thanks

githubnemo commented on src/peft/tuners/lora/layer.py:
```diff
@@ -156,6 +156,7 @@ def update_layer(
         arrow_config: ArrowConfig = None,
         qalora_group_size: int = 32,
         inference_mode: bool = False,
+        tied_adapters: Optional[dict[str, nn.Parameter]] = None,
```
Is this a misnomer? IIUC it only contains LoRA parameters for one adapter?

romitjain commented on src/peft/tuners/lora/model.py:
```diff
@@ -861,8 +870,63 @@ def subtract_mutated_init(self, output_state_dict: dict[str, torch.Tensor], adap
+        Args:
+            peft_config (LoraConfig)
+            tied_weight_keys (list[str])
+        """
```
I have updated the docstring

romitjain commented on src/peft/tuners/lora/model.py:
```diff
@@ -861,8 +870,63 @@ def subtract_mutated_init(self, output_state_dict: dict[str, torch.Tensor], adap
+
+        embed_layer_name = find_parameter_name_by_tensor(self.model, self.model.get_input_embeddings().weight)
+        # find_parameter_name_by_tensor returns the parameter name, so we need to strip the weight from the name
+        embed_layer_name = embed_layer_name.replace(".weight", "").replace("model.", "")
```
Updated the logic

romitjain commented on src/peft/tuners/lora/model.py:
```diff
@@ -861,8 +870,63 @@ def subtract_mutated_init(self, output_state_dict: dict[str, torch.Tensor], adap
+
+        for m in tied_weight_keys:
+            if m in modules_to_save:
+                modules_to_save.remove(m)
```
@githubnemo In the flow, I am unable to find any place where we are converting keys in `modules_to_save` to fully qualified keys.

The only two relevant checks I see are:
1. https://github.com/huggingface/peft/blob/main/src/peft/tuners/tuners_utils.py#L1655
2. https://github.com/huggingface/peft/blob/main/src/peft/utils/other.py#L1016


Here, we want to make sure that fully qualified keys from `tied_weight_keys` match the ones in `modules_to_save`. I propose that I do the following:

For every key in `model.named_parameters`, I perform a check similar to what is given in (1) and match it with `tied_weight_keys`. If both of them give a match, I remove the key from `modules_to_save`

romitjain commented:
Hi @githubnemo, gentle reminder for your review :)

BenjaminBossan commented:
@romitjain Sorry for the delay, we will hopefully get this finishes soon. As the corresponding PR for trainable tokens was merged, we now have some merge conflicts, but that should be easily resolved. Could you please take care?

romitjain commented:
@BenjaminBossan Resolved the merge conflict

romitjain commented on src/peft/tuners/lora/layer.py:
```diff
@@ -150,6 +150,19 @@ def update_layer(
         adapter_name: str,
         r: int,
         lora_alpha: int,
+        lora_dropout,
```
I think this was added by merge commit, let me take a look and remove this if not required

githubnemo commented on tests/test_initialization.py:
```diff
@@ -5064,6 +5067,128 @@ def test_weight_tying_tied_model_lokr(self):
+        assert embed_lora_A.data_ptr() == lm_lora_B.data_ptr()
+        assert embed_lora_B.data_ptr() == lm_lora_A.data_ptr()
+
+    @pytest.mark.parametrize("layer", [".*embed_tokens$", ".*lm_head$", ".*(embed_tokens|lm_head)$"])
```
Let's rename `layer` to `target_modules` in accordance with https://github.com/huggingface/peft/pull/2879/files#r2486161404

githubnemo commented on src/peft/tuners/tuners_utils.py:
```diff
@@ -1919,6 +1977,24 @@ def replicate_layers(model: nn.Module, layer_map: list[tuple[int, int]]):
+        if m is reference_tensor:
+            return n
+
+    return ""
```
I think this is a misnomer now. `reference_tensor` should actually be `reference_module`. Let's change the function name as well to match the new behavior.

githubnemo commented on src/peft/tuners/lora/model.py:
```diff
@@ -857,8 +867,86 @@ def subtract_mutated_init(self, output_state_dict: dict[str, torch.Tensor], adap
+        embed_layer_name = find_parameter_name_by_tensor(self.model, self.model.get_input_embeddings())
+        # find_parameter_name_by_tensor returns the parameter name, so we need to strip the weight from the name
+        if embed_layer_name.endswith(".weight"):
+            embed_layer_name = embed_layer_name.removesuffix(".weight")
```
Same comment as above

githubnemo commented on src/peft/tuners/lora/model.py:
```diff
@@ -857,8 +867,86 @@ def subtract_mutated_init(self, output_state_dict: dict[str, torch.Tensor], adap
+        if sep and "model" in prefix:
+            embed_layer_name = suffix
+
+        if isinstance(raw_target_modules, str):
```
I think that `target_modules="all-linear"` qualifies for this case (technically) but it probably doesn't reach this code path since it doesn't match the embedding layer. Still, I think we should have a test that makes sure that `LoraConfig(target_modules="all-linear", ensure_weight_tying=True)` doesn't break (if it doesn't exist already, I may have missed it!)

githubnemo reviewed (commented):
Hey @romitjain,

thanks for the updates and keeping up-to-date with main!

I have a few medium sized comments but generally this looks good.

One note regarding the review process: I think it's best to let the reviewer mark the issues as resolved to minimize the chances of missing a marked but still incomplete comment. It also helps me keep track of what I need to revise next time. It would help me a lot if we did it that way. Thanks :)

romitjain commented:
@githubnemo I have updated the PR to address your latest comments.

This is the commit - https://github.com/huggingface/peft/pull/2879/commits/19929f72f0eb7c07d39043f0ecb37a4f9b6f86a3 that addresses your comments.

Majorly:

1. Added tests for enc-dec (3 new tests), 1 test for all-linear. There are a total of 34 tests for weight tying across `modules_to_save` and `target_modules` (all passing locally)
2. I have done some small cleanup, expanded comments and removed artifacts of other merges or my older implmentation.

I know this PR has gone a few iterations over a long time. Thank you for reviewing it. I will be responisve on this PR, so appreciate your review.

romitjain commented:
@githubnemo gentle reminder

githubnemo reviewed (approved):
LGTM :)

Thanks for sticking with the PR and your continued effort!

Failing CI is unrelated.

romitjain commented:
Thanks @githubnemo and @BenjaminBossan for the continued review 🙂
